tests/utils/tensorflow/test_rasa_layers.py

```diff
@@ -0,0 +1,748 @@
+)
+realistic_feature_dense_seq_1 = tf.convert_to_tensor(
+    [[[10.0], [20.0], [30.0]], [[40.0], [50.0], [0.0]],], dtype=tf.float32
+)
+
+realistic_feature_signature_dense_2 = FeatureSignature(
+    is_sparse=False, units=2, number_of_dimensions=3
+)
+realistic_feature_dense_seq_2 = tf.convert_to_tensor(
+    [[[1.0, 2.0], [3.0, 4.0], [5.0, 6.0]], [[1.5, 2.5], [3.5, 4.5], [0.0, 0.0]],],
+    dtype=tf.float32,
+)
+
+realistic_feature_signature_dense_3 = FeatureSignature(
+    is_sparse=False, units=3, number_of_dimensions=3
+)
+realistic_feature_dense_sent_3 = tf.convert_to_tensor(
+    [[[0.1, 0.2, 0.3]], [[0.4, 0.5, 0.6]]], dtype=tf.float32,
+)
+
+realistic_sequence_lengths = tf.convert_to_tensor([3, 2], dtype=tf.int32)
+realistic_sequence_lengths_empty = tf.convert_to_tensor([0, 0], dtype=tf.int32)
+
+
+def test_concat_sparse_dense_correct_output_for_dense_input():
+    layer = ConcatenateSparseDenseFeatures(
+        attribute=attribute_name,
+        feature_type=SEQUENCE,
+        feature_type_signature=[
+            realistic_feature_signature_dense_1,
+            realistic_feature_signature_dense_2,
+        ],
+        config=dict(
+            model_config_basic,
+            **{SPARSE_INPUT_DROPOUT: True, DENSE_INPUT_DROPOUT: True},
+        ),  # also set all dropout to check that it has no effect on dense features
+    )
+    outputs_expected = [
+        [[10.0, 1.0, 2.0], [20.0, 3.0, 4.0], [30.0, 5.0, 6.0]],
+        [[40.0, 1.5, 2.5], [50.0, 3.5, 4.5], [0.0, 0.0, 0.0]],
+    ]
+    inputs = ([realistic_feature_dense_seq_1, realistic_feature_dense_seq_2],)
+    outputs = layer(inputs)
+    assert (outputs.numpy() == outputs_expected).all()
+
+
+def test_concat_sparse_dense_applies_dropout_to_sparse_input():
+    layer_dropout_for_sparse = ConcatenateSparseDenseFeatures(
+        attribute=attribute_name,
+        feature_type=SEQUENCE,
+        feature_type_signature=[feature_signature_sparse, feature_signature_sparse],
+        config=dict(model_config_basic, **{SPARSE_INPUT_DROPOUT: True, DROP_RATE: 1.0}),
+    )
+
+    inputs = ([feature_sparse_seq_3d, feature_sparse_seq_3d],)
+    outputs = layer_dropout_for_sparse(inputs, training=True)
+    expected_outputs = tf.zeros((batch_size, max_seq_length, units_sparse_to_dense * 2))
+
+    assert np.allclose(outputs.numpy(), expected_outputs.numpy())
+
+
+def test_concat_sparse_dense_applies_dropout_to_sparse_densified_input():
+    layer_dropout_for_sparse_densified = ConcatenateSparseDenseFeatures(
+        attribute=attribute_name,
+        feature_type=SEQUENCE,
+        feature_type_signature=[feature_signature_sparse, feature_signature_sparse],
+        config=dict(
+            model_config_basic, **{DENSE_INPUT_DROPOUT: True, DROP_RATE: 0.99999999}
+        ),  # keras dropout doesn't accept velues >= 1.0
+    )
+
+    inputs = ([feature_sparse_seq_3d, feature_sparse_seq_3d],)
+    outputs = layer_dropout_for_sparse_densified(inputs, training=True)
```

Comment:
please also check that if `training=False`, dropout is not applied
```suggestion
    train_outputs = layer_dropout_for_sparse_densified(inputs, training=True)
```

Reply:
Yep, now I'm check this too 🙂